File: 0/Templates/CSharp/ProjectTemplates/NavigationAppTemplate/Presentation/NavigationItemExtensions.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace $safeprojectname$.Presentation
{
    /// <summary>
    /// Extension helpers for the <see cref="NavigationItem"/> class.
    /// </summary>
    public static class NavigationItemExtensions
    {
        /// <summary>
        /// Returns a collection of ancestors of specified item.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public static IEnumerable<NavigationItem> GetAncestors(this NavigationItem item)
        {
            if (item == null) {
                throw new ArgumentNullException(nameof(item));
            }

            while (true) {
                item = item.Parent;
                if (item != null) {
                    yield return item;
                }
                else {
                    break;
                }
            }
        }

        /// <summary>
        /// Returns a collection of items that contain specified item, and the ancestors of specified item.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public static IEnumerable<NavigationItem> GetAncestorsAndSelf(this NavigationItem item)
        {
            if (item == null) {
                throw new ArgumentNullException(nameof(item));
            }

            while (true) {
                if (item != null) {
                    yield return item;
                }
                else {
                    break;
                }

                item = item.Parent;
            }
        }

        /// <summary>
        /// Retrieves the descendant child items of specified item.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public static IEnumerable<NavigationItem> GetDescendants(this NavigationItem item)
        {
            if (item == null) {
                throw new ArgumentNullException(nameof(item));
            }

            var stack = new Stack<NavigationItem>(item.Items.Reverse());
            return GetDescendantsAndSelf(stack);
        }

        /// <summary>
        /// Returns a collection of items that contain the specified item and all descendant items.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public static IEnumerable<NavigationItem> GetDescendantsAndSelf(this NavigationItem item)
        {
            if (item == null) {
                throw new ArgumentNullException(nameof(item));
            }

            var stack = new Stack<NavigationItem>();
            stack.Push(item);
            return GetDescendantsAndSelf(stack);
        }

        private static IEnumerable<NavigationItem> GetDescendantsAndSelf(Stack<NavigationItem> stack)
        {
            while (stack.Count > 0) {
                var item = stack.Pop();
                yield return item;

                foreach (var child in item.Items.Reverse()) {
                    stack.Push(child);
                }
            }
        }

        /// <summary>
        /// Determines whether specified item is the root item, that is it has no parent.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public static bool IsRoot(this NavigationItem item)
        {
            if (item == null) {
                throw new ArgumentNullException(nameof(item));
            }

            return item.Parent == null;
        }

        /// <summary>
        /// Determines whether the item is a leaf item, meaning it has no children.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public static bool IsLeaf(this NavigationItem item)
        {
            if (item == null) {
                throw new ArgumentNullException(nameof(item));
            }

            return !item.Items.Any();
        }

        /// <summary>
        /// Determines whether the item has grandchildren.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public static bool HasGrandchildren(this NavigationItem item)
        {
            if (item == null) {
                throw new ArgumentNullException(nameof(item));
            }

            return item.Items.Any(i => !i.IsLeaf());
        }
    }
}

File: 0/Templates/VisualBasic/ProjectTemplates/NavigationAppVBTemplate/Presentation/NavigationItemExtensions.vb
Namespace Presentation
    Module NavigationItemExtensions

        ''' <summary>
        ''' Returns a collection of items that contain specified item, and the ancestors of specified item.
        ''' </summary>
        ''' <param name="item"></param>
        ''' <returns></returns>
        <Extension>
        Public Function GetAncestorsAndSelf(item As NavigationItem) As IEnumerable(Of NavigationItem)
            If item Is Nothing Then Throw New ArgumentException(NameOf(item))

            Dim result As List(Of NavigationItem) = New List(Of NavigationItem)

            While True
                If item Is Nothing Then Exit While
                result.Add(item)
                item = item.Parent
            End While

            Return result
        End Function

        ''' <summary>
        ''' Determines whether specified item is the root item, that is it has no parent.
        ''' </summary>
        ''' <param name="item"></param>
        ''' <returns></returns>
        <Extension>
        Public Function IsRoot(item As NavigationItem) As Boolean
            If item Is Nothing Then Throw New ArgumentException(NameOf(item))

            Return item.Parent Is Nothing
        End Function

        ''' <summary>
        ''' Determines whether the item is a leaf item, meaning it has no children.
        ''' </summary>
        ''' <param name="item"></param>
        ''' <returns></returns>
        <Extension>
        Public Function IsLeaf(item As NavigationItem) As Boolean
            If item Is Nothing Then Throw New ArgumentException(NameOf(item))

            Return Not item.Items.Any()
        End Function

        ''' <summary>
        ''' Determines whether the item has grandchildren.
        ''' </summary>
        ''' <param name="item"></param>
        ''' <returns></returns>
        <Extension>
        Public Function HasGrandchildren(item As NavigationItem) As Boolean
            If item Is Nothing Then Throw New ArgumentException(NameOf(item))

            Return item.Items.Any(Function(i) Not i.IsLeaf())
        End Function

    End Module
End Namespace

